### ISSUE TITLE
Add SSL_writev and SSL_writev_ex

### ISSUE THREAD

**Issue by peppapighs** [openssl/openssl#20918]
I would like to implement `SSL_writev()` and `SSL_writev_ex()` to support writing vectored data in a single record. 

As discussed [here](https://stackoverflow.com/questions/38198638/openssl-ssl-write-from-multiple-buffers-ssl-writev), the current solution would be having the user consolidates the data manually then call `SSL_write()` on the consolidated data.

Therefore, I propose adding two new APIs to combine this process:
```c
int SSL_writev(SSL *s, const struct iovec *iov, int iovcnt);
int SSL_writev_ex(SSL *s, const struct iovec *iov, size_t iovcnt, size_t *written);
```
It will require adding `ssl_writev()` and `ssl_writev_bytes` in `struct ssl_method_st`. `SSL_writev` then can make a call to `ssl_writev_internal` where it calls `ssl3_writev`.

Depending on the negotiated protocols, `ssl3_writev` then may call `ssl3_writev_bytes` or `dtls1_writev_app_data_bytes`.

In `ssl3_writev_bytes`, for each pipelines, we can allocate a temporary buffer with the size being the length of data we want to send through the pipeline. We then fill the buffer with the vectored data, possibly using `memcpy`. By constructing a temporary buffer, we can use the existing `write_records` call.

In `dtls1_writev_app_data_bytes`, we can also allocate a temporary buffer with the size being the total length of the vectored data then copy everything over and the subsequent calls should be the same.

**Comment by paulidale** [comment]
I'm wondering why this is required?  writev exists to avoid copying data, however the encryption step in TLS is a copy step.
I.e. why not call SSL_write multiple times?

Do you have any performance figures indicating that this is a significant gain?

**Comment by peppapighs** [comment]
The goal here is to reduce the overhead of sending multiple records when the payloads are small enough to be combined into a single record. Also, from the API perspective, many TCP libraries supports scatter/gather I/O so I think it is reasonable to implement such APIs here as well.

**Comment by paulidale** [comment]
I thought we coalesced records already.  @mattcaswell ?
I'm pretty sure there is an option to force it.

**Comment by peppapighs** [comment]
> I thought we coalesced records already. @mattcaswell ? I'm pretty sure there is an option to force it.
> 
> I'm not against having this API, it's just that it means an ongoing maintenance burden supporting it which means justifying it. The QUIC work is adding something similar.

Should I include the implementation of QUIC as well? I think we can implement something similar to `ossl_quic_sstream_append` that pushes the vectors in to the ring buffer.

**Comment by peppapighs** [comment]
> We do not have single copy public API for QUIC yet. The internals are mostly written with single copy in mind but there are things missing.
> 
> And this API SSL_writev is not the right api because you would need the SSL object to take over the buffers until it stops needing them.

So in the case of QUIC protocol, what should be the expected behavior for `SSL_writev()`? More specifically, in `ssl_writev_internal`, should I combine the vectors and call `ssl_write()`? or throw an error?

**Comment by t8m** [comment]
Neither of these would be acceptable. IMO converting it to individiual ossl_sstream() calls would be acceptable, except it still does not avoid copying the data into the ring buffer.

It is unclear if this API is something we want, adding OTC hold.

**Comment by peppapighs** [comment]
https://github.com/peppapighs/ssl-writev-benchmark/blob/implementation/README.md

Here is the benchmark result showing the performance of this `SSL_writev()` implementation. Hopefully, this will contribute to your decision.

**Comment by paulidale** [comment]
Great work.  That's a larger improvement than I expected.

**Comment by peppapighs** [comment]
There is a bug in my benchmark result where the statistics of `SSL_writev()` is higher than usual (the previous result of `SSL_writev()` was calculated from data points of `SSL_writev()` combined with `SSL_write()`). This has been fixed.

I also included the execution time of combining vectors into a single buffer and calling `SSL_write()` to serve as the baseline performance. The updated result can be viewed in the same README file.

**Comment by t8m** [comment]
Thank you for the benchmark results.

Not surprisingly the results for SSL_writev() and SSL_write() with coalescing the buffer before the call do not differ significantly. IMO that makes the SSL_writev() call rather a convenience function. This does not preclude adding it but perhaps some API that has single-copy semantics in mind (i.e. scatter/gather) would be a better idea.

**Comment by peppapighs** [comment]
The implementation right now saves a single copy when compared to coalescing method. If you have any other ideas on how we can implement this API, I am willing to help implement them.

**Comment by t8m** [comment]
It is not a problem with the implementation. It is a problem with how QUIC does resending frames when packets are lost. Basically this API won't fit its model.

**Comment by peppapighs** [comment]
May I know what is the plan for this issue? Are we going to wait for QUIC implementation to be complete before we revisit it?

**Comment by arbv** [comment]
Yes, please add one. It is quite a common thing to assemble data from multiple places before sending it via TLS. Now assembling all data in one buffer is required to avoid excessive TLS records, and often that implies extra memory copying. `SSL_weitev()` would have solved this one nicely.

**Comment by user202729** [comment]
Can I ask what's the status of this?

Looking at the comment under the pull request, the 2 blocking points are

> There could be other ways to implement this without a new API, for example by allowing libssl to maintain an in-progress unfinished record across multiple SSL_write calls. We would need a suitable API to force flushing; possibly one already exists(?).

Wouldn't this break backwards compatibility (the new `SSL_write` function does not always flush, so code that relies on this behavior will become incorrect)?
But otherwise sounds reasonable?

> This API creates a dependency on struct iovec, which is not available on all platforms and thus cannot be merged as is. It would need revision with portability in mind. We have already created our own internal iovec-like structures for internal use with our QUIC code.

You mean OSSL_QTX_IOVEC? Maybe we can expose the structure and consolidate the implementation?

Although strictly speaking the linked pull request does not create a dependency on struct iovec (it defines struct iovec independently if it's not available on a given platform). We could remove the guarantee in the API that struct iovec is identical to that of the operating system, or make an alias under a different name.

**Comment by nhorman** [comment]
To answer your immediate question, this issue is unfortunately stalled for lack of time to address it.  Its marked for consideration in a future release, but that is not a time guarantee.  If you have interest in writing a PR for it, I'm happy to assign it to you as a community issue.

**Comment by user202729** [comment]
Yes, I'm interested in writing a PR.

But what design of the API do you prefer (considering that the previous pull request are not merged)?

Apart from writev, another possibility of implementing something like SSL_cork() … SL_uncork() which buffers every SSL_write() calls inside.

**Comment by nhorman** [comment]
we should have some better guidance once OTC discusses this

**Comment by dridi** [comment]
More thoughts on this, I wasn't particularly concerned with this change but one remark seems to have stuck in my mind, the disqualifying idea that `SSL_writev()` would be a mere convenience on top of `SSL_write()`.

I hope that my poll-once-write-once remark is in itself an argument against the idea of the function being just a convenience.

The question of portability is near and dear to me so I can only sympathize with this counter-argument.

I'd like to postulate the idea of `SSL_write()` becoming the convenience and `SSL_writev()` actually becoming the norm:

```c
#ifdef OPENSSL_NO_IOVEC
typedef struct {
	void   *iov_base;
	size_t  iov_len;
} SSL_iovec;
#else
typedef struct iovec SSL_iovec;
#endif

int
SSL_writev(SSL *ssl, const SSL_iovec *iov, int num)
{

	/* current write logic, feeding from multiple buffers at once */
}

int
SSL_write(SSL *ssl, const void *buf, int num)
{
	SSL_iovec iov;

	iov.iov_base = (void *)buf;
	iov.iov_len = num;
	return (SSL_writev(ssl, &iov, 1));
}
```

On targets that provide a `struct iovec` definition you get a proper integration, and otherwise a really thin compatibility layer. It's up to the `SSL_writev()` caller to deal with partial writes just like a regular `writev()` so it shouldn't complicate the current logic beyond iterating over buffers.

**Comment by alfaloo** [comment]
I concur with the points made by @dridi . Specifically, I believe that the use case of SSL_writev is compelling enough that we can consider replacing/updating all those internal write calls to work for ssl_writev, then just change ssl_write to be a simplified wrapper to ssl_writev. the performance of the ssl_write use case will not be affected as we are just additionally iterating through the array of iovec structures.

with regards to the problem of iovec not available on all platforms, we can create a compatibility layer that abstracts away the differences.

i am interested in writing a pr for this proposed implementation. i'd love to hear some thoughts from everyone else
